### ISSUE TITLE
Question: Code for computing angle of attack and glider speed

### ISSUE THREAD

**Issue by hawesie** [smerckel/gliderflight#2]
Thanks again for all your amazing open-source code!

My use case is building a very low-fidelity simulator of a Slocum glider to (quickly and approximately) predict surfacing locations given a starting GPS point and a mission file. I have a calibrated glider model using `gliderflight` (using dbd files from historical missions of the same glider), and now I'd like to calculate the angle of attack for the dive and the vertical and horizontal speeds (which are the main additional parameters of my simulator). 

Looking at your 2019 paper it seems like I should first solve equation 9 to get the angle of attack (using the commanded pitch angle plus the values `fm.Cd0`, `fm.Cd1`, `fm.ah` from the glider model), then calculate values for F_g and F_B (equations, 3 and 4, using `fm.epsilon` etc), and finally solve equation 8 for U (which I can then project to x, and z). Does that sound correct? If so, I was wondering if any of the functions in `gliderflight` already do parts of this in a way that I can reuse?

**Comment by hawesie** [comment]
Looks like I can answer part of my question already: `SteadyStateGliderModel.solve_for_angle_of_attack` appears to give me back the angle of attack for a given pitch value. 

Related bug report: the docs says `SteadyStateGliderModel.solve_for_angle_of_attack` can accept a scalar, but it seems to need a `np.array`).

**Comment by hawesie** [comment]
I think I've now got the answer.  I can reuse the following from `SteadyStateGliderModel.solve`. 

```
pressure, Vbp = self.convert_pressure_Vbp_to_SI(data['pressure'],
                                                        data['buoyancy_change'])
FB, Fg = self.compute_FB_and_Fg(pressure, rho, Vbp)
alpha, U = self.solve_model(rho, FB, pitch, Fg)
wg = np.sin(pitch+alpha)*U
ug = np.cos(pitch+alpha)*U
```

Since I'm doing this ahead of the dive I don't have `pressure` or `density` but I can use an approximate value for a depth in the middle of the dive, and I assume that will be fine for my purposes.

**Comment by smerckel** [comment]
Sure, your use case can reuse methods. You would need to setup a SteadyStateGliderModel, and at least define Cd0, mg and Vg. The model is designed to deal with input data provided as numpy arrays, but it also deals with scalar version. In the example below, I set some input data. The density is as measured, or, most likely in your case modelled using some GCM. Note that this is the in-situ density, not the potential density, which is most likely the model output. Other parameters are in the units as used by the Slocum gliders. Pressure and buoyancy_change need to be converted into SI units. I have not really intended to use the class like this, but if U and alpha are the only parameters you want, it is easy to  do, using the solve_model() method, as shown below. If you also want to have the horizontal and vertical velocities, you'd need to use the method solve(), and prepare a dictionary with data. The method assumes dhdt to be present, if not if will try to compute that from time and pressure. With scalars this fails, so you can set dhdt to 0, and time as well. The output is a named tuple that gives in addition to U and alpha, also the attributes to u and w for the horizontal and vertical velocities, respectively. The code below should clarify this, I hope.

````
import gliderflight as gf

# Input data
rho = 1025.3 # in-situ density of water
pitch = 0.433
pressure = 3 # bar
buoyancy_change = 233 # cc

# construct a data dictionary for the solve() method, see below.
data = dict(pitch=pitch, density=rho, pressure=pressure,
            buoyancy_change=buoyancy_change, time=0, dhdt=0)

# set up the model

gm = gf.SteadyStateGliderModel()
gm.define(Cd0=0.12, mg=65, Vg=65/1025.)


# The model assumes SI units, so convert pressure and buoyancy change
# values accordingly.
pressure, buoyancy_change = gm.convert_pressure_Vbp_to_SI(pressure,
                                                          buoyancy_change)

FB, Fg = gm.compute_FB_and_Fg(pressure, rho, buoyancy_change)
alpha, U = gm.solve_model(rho, FB, pitch, Fg)

# If you want to compute the horizontal and vertical velocity
# components directly, then you would want to call the solve() method.


model_result = gm.solve(data)

# Be certain that both methods give the same result.
assert alpha == model_result.alpha
assert U == model_result.U

# assign horizontal and vertical glider speeds:
horizontal_glider_speed = model_result.u
vertical_glider_speed = model_result.w
````

**Comment by hawesie** [comment]
Amazing, thanks!

**Comment by hawesie** [comment]
To make your code work I need to set pitch as an `np.array`

```python
pitch = np.array([0.433])
```

**Comment by smerckel** [comment]
I guess you don't use the latest github version? Version 1.0.1 is the version as used in the 2019 paper. If you prefer to install from pypi I can push it there. Main difference is the parallel computing of the dynamic flight model and some small things.

**Comment by smerckel** [comment]
Published new version 1.2.0 on pypi.
